File: src/pablo/providers/linear.py
"""Linear provider, backed by schpet's ``linear`` CLI.

Exact subcommands used (documented in @README.md; the CLI was chosen but not
yet installed on this machine — verify these against ``linear --help`` on
first install and adjust here if they differ):
- ``linear issue view <KEY> --json``
- ``linear issue list --assignee <identity> --json``
- ``linear auth status``  — auth check
"""

from __future__ import annotations

import json
import re
from datetime import datetime

from pablo.config import ProjectConfig
from pablo.model import Issue, Task
from pablo.providers import parse_ts, run_cli

_ISSUE_URL_RE = re.compile(r"https?://linear\.app/[^/]+/issue/([A-Z][A-Z0-9]*-\d+)")


class LinearProvider:
    name = "linear"

    def match_url(self, url: str, cfg: ProjectConfig) -> str | None:
        match = _ISSUE_URL_RE.match(url)
        if not match:
            return None
        key = match.group(1)
        if key.split("-")[0] != (cfg.project_key or ""):
            return None
        return key

    def _view(self, key: str) -> dict:
        return json.loads(run_cli(["linear", "issue", "view", key, "--json"]))

    def _issue_from_json(self, data: dict, cfg: ProjectConfig) -> Issue:
        return Issue(
            provider=self.name,
            key=data["identifier"],
            url=data.get("url", ""),
            title=data["title"],
            project_key=cfg.project_key,
            status=(data.get("state") or {}).get("name"),
        )

    def get_issue(self, ref: str, cfg: ProjectConfig) -> Issue:
        return self._issue_from_json(self._view(ref), cfg)

    def list_assigned(self, cfg: ProjectConfig) -> list[Issue]:
        out = run_cli(
            ["linear", "issue", "list", "--assignee", cfg.identity, "--json"]
        )
        return [self._issue_from_json(item, cfg) for item in json.loads(out)]

    def issue_status(self, key: str, cfg: ProjectConfig) -> str:
        return self._view(key)["state"]["name"]

    def failure_signal_events(self, task: Task, cfg: ProjectConfig) -> list[datetime]:
        """Timestamps of history transitions into ``testing.failure_signal``."""
        if task.issue is None or not cfg.failure_signal:
            return []
        data = self._view(task.issue.key)
        stamps = [
            parse_ts(entry["createdAt"])
            for entry in data.get("history", [])
            if (entry.get("toState") or {}).get("name") == cfg.failure_signal
        ]
        return sorted(stamps)

    def cli_name(self) -> str:
        return "linear"

    def auth_check_cmd(self) -> list[str]:
        return ["linear", "auth", "status"]

File: app/src/Provider/Tracker/Linear.php
<?php

declare(strict_types=1);

namespace Pablo\Provider\Tracker;

use Pablo\Config\ProjectConfig;
use Pablo\Domain\Issue;
use Pablo\Domain\Task;
use Pablo\Support\Proc;

/**
 * Linear provider, backed by schpet's linear CLI.
 */
final class Linear implements Provider
{
    public function name(): string
    {
        return 'linear';
    }

    public function supportsSignalViaStatus(): bool
    {
        return false;
    }

    public function matchUrl(string $url, ProjectConfig $cfg): ?string
    {
        if (1 !== preg_match('#https?://linear\.app/[^/]+/issue/([A-Z][A-Z0-9]*-\d+)#', $url, $m)) {
            return null;
        }
        $key = $m[1];
        if (explode('-', $key)[0] !== $cfg->projectKey) {
            return null;
        }

        return $key;
    }

    /** @return array<string, mixed> */
    private function view(string $key): array
    {
        $out = Proc::run(['linear', 'issue', 'view', $key, '--json']);
        /** @var array<string, mixed> $data */
        $data = json_decode($out, true);

        return $data;
    }

    /**
     * @param array<string, mixed> $data
     */
    private function issueFromJson(array $data, ProjectConfig $cfg): Issue
    {
        return new Issue(
            provider: $this->name(),
            key: (string) $data['identifier'],
            url: (string) ($data['url'] ?? ''),
            title: (string) $data['title'],
            projectKey: $cfg->projectKey,
            status: $data['state']['name'] ?? null,
        );
    }

    public function getIssue(string $ref, ProjectConfig $cfg): Issue
    {
        return $this->issueFromJson($this->view($ref), $cfg);
    }

    public function listAssigned(ProjectConfig $cfg): array
    {
        $out = Proc::run(['linear', 'issue', 'list', '--assignee', $cfg->identity, '--json']);
        $issues = [];
        /** @var array<int, array<string, mixed>> $items */
        $items = json_decode($out, true);
        foreach ($items as $item) {
            $issues[] = $this->issueFromJson($item, $cfg);
        }

        return $issues;
    }

    public function issueStatus(string $key, ProjectConfig $cfg): string
    {
        $data = $this->view($key);

        return (string) $data['state']['name'];
    }

    public function batchIssueStatus(array $pairs): array
    {
        if ([] === $pairs) {
            return [];
        }

        $commands = [];
        $keyIndex = [];
        foreach ($pairs as $i => [$key]) {
            $keyIndex[$i] = $key;
            $commands[] = ['linear', 'issue', 'view', $key, '--json'];
        }

        $results = Proc::runParallel($commands, check: false);
        $statuses = [];
        foreach ($results as $i => $out) {
            $key = $keyIndex[$i];
            try {
                /** @var array<string, mixed> $data */
                $data = json_decode($out, true, 512, \JSON_THROW_ON_ERROR);
                $statuses[$key] = (string) $data['state']['name'];
            } catch (\Throwable) {
                $statuses[$key] = '?';
            }
        }

        return $statuses;
    }

    public function failureSignalEvents(Task $task, ProjectConfig $cfg): array
    {
        if (null === $task->issue || null === $cfg->failureSignal) {
            return [];
        }
        $data = $this->view($task->issue->key);
        $stamps = [];
        foreach ($data['history'] ?? [] as $entry) {
            if (($entry['toState']['name'] ?? null) === $cfg->failureSignal) {
                $stamps[] = Proc::parseTs((string) $entry['createdAt']);
            }
        }
        sort($stamps);

        return $stamps;
    }

    public function cliName(): string
    {
        return 'linear';
    }

    public function authCheckCmd(): array
    {
        return ['linear', 'auth', 'status'];
    }
}
